### ISSUE TITLE
Plans for Community Typescript Extensions Repo

### ISSUE THREAD

**Issue by ricklove** [buehler/typescript-hero#174]
This is great! 

But I think we all need to join forces and make it even better!

There are a few patterns I have found myself repeating often in typescript and I missed some of the refactoring tools I had in Visual Studio with C#. 

I was thinking about creating my own extension to speed up my own coding, but of course I wanted to find existing projects first.

Here are a few I have found:

- Typescript Hero 
    - Intellisense with missing Imports - Nice!
    - Auto Imports - Excellent!
- [Vscode-Refactorix](https://github.com/krizzdewizz/vscode-refactorix)
    - A few short tools
- [Vscode-GenGetSet](https://github.com/cybertim/vscode-gengetset)
    - Looks like it also has Auto Imports and some other stuff
- [Js-Refactor](https://github.com/cmstead/js-refactor)
    - Some good looking method management that I would like to see in Typescript (Similar to Generate Method from Visual Studio/C#)
- [Resharper Visual Studio Typescript](https://www.jetbrains.com/help/resharper/2016.3/ReSharper_by_Language__TypeScript__Refactorings.html)
    - A few interesting refactorings like introduce field
- [Atom Editor Typescript Refactorings](https://github.com/TypeStrong/atom-typescript/blob/master/docs/quickfix.md)
    - Add class members and some handle refactorings

Anyway, I'm sure there are others. 

My main point is that I am interesting in partnering with you and others in creating a community repo for typescript refactoring. I imagine that a community organization would be created with shared ownership so that multiple owners can help manage pull requests, etc.

I like your branding with Typescript Hero, and I think if you moved it to a GitHub organization and allowed others to join in, it could be the goto tool for Typescript Refactoring.

Let me know what you think.

**Comment by buehler** [comment]
Hey @ricklove 

Basically I love the idea! I would really love to see a cool developer experience grow from all those different projects. Actually with TypeScript 2.2, the whole auto import thing will kind of die, since they introduced some things that I implemented. But the good thing is, we could start add cool stuff.

As of now, some things are still missing in the typescript server; when one types, no code completion is provided. You have to write the class name and can then import it automatically via a code action.

I'm currently rewriting the whole extension to use client / server paradigm that is purposed by microsoft, so if this could be a possibility, I'd like to join forces after this rewrite.

Basically we could "just disable" the auto import stuff and focus on the cool ideas (like refactorings, automatic class generation, etc)

Did you open a ticket in the other repos you mentioned? I'd be more than happy to share the whole thing!

Cheers
Chris

**Comment by ricklove** [comment]
Ok, cool. 

That would be great to see what you think is an ideal architecture.

I was thinking that a higher level architecture could be created that would allow the same refactorings to work in vscode and other editors (abstracting away specifics of each editor and building a set of refactoring utilities). 

Also, it would be cool if there is something like the json snippet editor for refactorings. Json alone wouldn't be enough, but a plugin architecture where users can easily make their own local refactorings would be great. Possibly, this could be coupled with a quick and easy way to submit those to the community.

- What are your thoughts about the architecture?

I have not opened tickets in the other repos, I thought it would be easier to figure out a small plan then reach out to everyone.

Perhaps once a community organization is setup then it would be a good time to invite everyone to join a discussion there.

- Do you think that would be a good way forward?

**Comment by buehler** [comment]
Hey @ricklove 

This could get awesome :-) Well IMO the first step would be to kind of split up typescript hero. So we actually could make a nodejs-package out of the whole parsing / indexing part (especially the TypescriptParser and the reverse-index of the declarations). With that package, we could think about abstraction layers (interfaces) that could abstract an editor und would need to be implemented for each one. So basically the core of ts-hero would be the calculation and parsing of the files and the delivery of the declarations for certain use-cases.

As an example: we could have a `TypescriptHeroAutoComplete` which contains methods that helps deliver intellisense and actual declarations used for intellisense.

With such a (maybe second) package, the implementation should be - at least more or less - easy for other editors to come by. We could then even provide the implementations for other editors.

I totally agree with you that we need a small plan to get everyones attention.

Another thing could be to start and add another parser (ES6) to actually enable people to use our refactoring / importing tools with ES6 and not only typescript (at least far down the road).

**Comment by ricklove** [comment]
Nice. Typescript Heroes is perfect.

I hadn't thought of ES6. Hopefully since Typescript is just a superset of ES, it won't be hard to use the typescript toolchain for both.

In fact you just gave an idea for some ultimate refactorings:

Upgrade to Typescript (ES5/6->Typescript ES7):

- Typify everything
- Flatten promises/then/catch -> async/await/try/catch
- Upgrade ES5 code to this era (let, const, lamda, class, import, etc.)

Also the reverse could be useful for working with the anti-Typescript (for example Typescript developers who are working with ES6 teams).

Downgrade from Typescript: (TS->ES6)

- Just run a build on the current file or selection (instead of building the entire project with tsc)

Generate Local .d.ts

- Generate a simple and editable .d.ts for untyped modules.
- Possible choose the components of concern and generate the definitions for only those components.


Anyway, these are just some ideas that came to mind that would be helpful.

Of course they would be a major undertaking, but it could start with local refactorings that work on the method level and as they work better and all cases are handled it could eventually take on a whole file or more.

I look forward to having a good architecture where it can be simple to build next-gen refactorings and greater increase coding speed.

**Comment by buehler** [comment]
Cool ideas! 👍 

Actually, I just tried, we can throw normal javascript (ES5/6) at the typescript parser, and it will generate the correct AST and declarations. So basically, a first step to a cool architecture would be to split the core (parsing / reverse index) of tsh out and create a separate package. With that implement typescript hero for vscode and possible other extensions for other editors or languages like ES6.

After that step, I'd recommend to start abstract the different ways of refactoring. Like fix code by implementing a method or something like that. We need to provide the helpers for it, but not whole implementations, that's a topic for the concrete extensions (for sublime or whatever)

So we agree on making a community and try to harmonize our efforts.

First steps would be that I finish the rewrite, so the extension is working as before again and then we can split the core parts out.

**Comment by ricklove** [comment]
Yeah, sounds good. 

How long do you think it will take you to finish the rewrite?

In the mean time, it might make sense to begin a discussion in the community and start inviting people to that discussion

**Comment by buehler** [comment]
Well, depends heavily on the spare time I got to work on it. But at least two weeks or so. I'm nearly at feature parity but well :-) it needs time.

For the other topic, absolutely! We should start invent the creators of the other extensions and bring them on board.

**Comment by buehler** [comment]
Alright, yesterday I was @ microsoft (switzerland) and had a cool talk with johannes rieken (developer with microsoft)

Actually there is a plan to get some extensions into the ts-server (the part that parses the files etc.) much like you can write extensions for the roslyn compiler.

So instead of pulling my hair of with further rewriting the whole extension, I'll wait for the ts-server part and try to hook into the ts server after they released the feature.

This will produce a much more cleaner extension since not both extensions need to parse everything.

(fyi: @ricklove )
